File: src/main/webapp/JS/validation.js
// Element IDs
const USER_NAME_ERROR = "userNameError";
const EMAIL_ERROR = "emailError";
const PASSWORD_ERROR = "passwordError";
const RE_PASSWORD_ERROR = "rePasswordError";
const STAFF_ID_ERROR = "staffIdError";
const ROLE_ERROR = "roleError";
const LOCATION_FORM = "locationForm";
const LOCATION_ERROR = "locationError";
const THEATRE_ID = "theatreId";
const THEATRE_NAME_ERROR = "theatreNameError";
const ZIP_ERROR = "zipError";
const ROOM_ID = "roomId";
const ROOM_NUMBER_ERROR = "roomNumberError";
const SECTIONS_ERROR = "sectionsError";
const SEATS_ERROR = "seatsError";
const ERROR_MESSAGE = "errorMessage";

// Ajax URLs
const FIND_REGISTERED_EMAIL = "FindRegisteredEmail";
const FIND_THEATRE_NAME = "FindTheatreName";
const FIND_ROOM_NUMBER = "FindRoomNumber";

// Input limits
const USERNAME_MIN_LENGTH = 2;
const USER_NAME_MAX_LENGTH = 20;
const PASSWORD_MIN_LENGTH = 8;


// Validate customer signup form
function validateSignUp(form) {
    fail = checkName(form.user_name);
    checkEmail(form.email).then(invalidInput => {
        fail += invalidInput;
        if (fail == "") {
            document.getElementById(form.id).submit();
        }
    });
    fail += checkPassword(form.password);
    fail += checkRePassword(form.id);
    return false;
}

// Validate staff signup form
function validateStaffSignUp(form, isAdmin) {
    fail = "";
    if (isAdmin) {
        fail += checkRole(form.role);
        invalidLocation = "";
        if (form.role.value != "admin") {
            fail += checkLocation(form.theatre_location);
        }
    }
    fail += checkName(form.user_name);
    checkEmail(form.email).then(invalidInput => {
        fail += invalidInput;
        if (fail == "") {
            document.getElementById(form.id).submit();
        }
    });
    fail += checkPassword(form.password);
    return false;
}

// Validate customer signin form
function validateSignIn(form) {
    fail = checkSignInEmail(form.email);
    fail += checkSignInPassword(form.password);
    return fail == "";
}

// Validate staff signin form
function validateStaffSignIn(form) {
    fail = checkSignInStaffId(form.staff_id);
    fail += checkSignInPassword(form.password);
    return fail == "";
}

// Validate theatre form
function validateTheatreForm(form) {
    errorId = document.getElementById(ERROR_MESSAGE);
    errorId.innerHTML = fail = "";
    if (form.theatre_name.value == "" || form.address.value == ""
        || form.city.value == "" || form.state.value == ""
        || form.country.value == "" || form.zip.value == "") {
        errorId.innerHTML = fail += "* required fields\n";
    }
    if (form.theatre_name.value != "") {
        checkTheatreName(form.theatre_name).then(invalidInput => {
            fail += invalidInput;
            if (fail == "") {
                document.getElementById(form.id).submit();
            }
        });
    }
    if (form.zip.value != "") {
        fail += checkZip(form.zip);
    }
    return false;
}

// Validate ticket price form
function validateTicketPriceForm(form) {
    errorId = document.getElementById(ERROR_MESSAGE);
    errorId.innerHTML = "";
    if (form.start_time.value == "") {
        errorId.innerHTML = "Please enter start time\n";
        return false;
    }
    if (form.price.value == "") {
        errorId.innerHTML = "Please enter price\n";
        return false;
    }
    return true;
}

// Validate room form
function validateRoomForm(form) {
    errorId = document.getElementById(ERROR_MESSAGE);
    errorId.innerHTML = fail = "";
    if (form.room_number.value == "" || form.sections.value == "" || form.seats.value == "") {
        errorId.innerHTML = fail += "* required fields\n";
    }
    if (form.room_number.value != "") {
        checkRoomNumber(form.room_number).then(invalidInput => {
            fail += invalidInput;
            if (fail == "") {
                document.getElementById(form.id).submit();
            }
        });
    }
    if (form.sections.value != "") {
        fail += checkSections(form.sections);
    }
    if (form.seats.value != "") {
        fail += checkSeats(form.seats);
    }
    return false;
}

// Validate movie form
function validateMovieForm(form) {
    errorId = document.getElementById(ERROR_MESSAGE);
    errorId.innerHTML = "";
    if (form.title.value == "" || form.release_date.value == ""
        || form.duration.value == "" || form.trailer.value == "" || form.description.value == "") {
        errorId.innerHTML = "* required fields\n";
        return false;
    }
    if (validateNumber(form.duration.value) != "") {
        errorId.innerHTML = "Invalid duration\n";
        return false;
    }
    return true;
}

// Validate schedule form
function validateScheduleForm(form) {
    errorId = document.getElementById(ERROR_MESSAGE);
    errorId.innerHTML = "";
    if (form.show_date.value == "") {
        errorId.innerHTML = "Please enter show date\n";
        return false;
    }
    if (form.start_time.value == "") {
        errorId.innerHTML = "Please enter start time\n";
        return false;
    }
    if (form.room_number.value == "") {
        errorId.innerHTML = "Please select a room\n";
        return false;
    }
    return true;
}

function checkName(nameInput) {
    invalidInput = validateUserName(nameInput.value);
    document.getElementById(USER_NAME_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

async function checkEmail(emailInput) {
    invalidInput = validateEmail(emailInput.value);
    if (invalidInput == "") {
        args = emailInput.name + "=" + emailInput.value;
        invalidInput = await ajaxFunction(args, FIND_REGISTERED_EMAIL);
    }
    document.getElementById(EMAIL_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

function checkPassword(passwordInput) {
    invalidInput = validatePassword(passwordInput.value)
    document.getElementById(PASSWORD_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

function checkRePassword(formId) {
    form = document.getElementById(formId);
    invalidInput = validateRePassword(form.password.value, form.re_password.value);
    document.getElementById(RE_PASSWORD_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

function checkRole(roleInput) {
    invalidInput = validateRole(roleInput.value)
    document.getElementById(ROLE_ERROR).innerHTML = invalidInput;
    if (invalidInput == "" && roleInput.value == "admin") {
        document.getElementById(LOCATION_FORM).setAttribute("hidden", "");
    } else {
        document.getElementById(LOCATION_FORM).removeAttribute("hidden");
    }
    return invalidInput;
}

function checkLocation(locationInput) {
    invalidInput = "";
    if (locationInput.value == "") {
        invalidInput = "Please select a theatre location\n";
    }
    document.getElementById(LOCATION_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

function checkSignInEmail(emailInput) {
    invalidInput = "";
    if (emailInput.value == "") {
        invalidInput = "Please enter your email\n";
    }
    document.getElementById(EMAIL_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

function checkSignInPassword(passwordInput) {
    invalidInput = "";
    if (passwordInput.value == "") {
        invalidInput = "Please enter your password\n";
    }
    document.getElementById(PASSWORD_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

function checkSignInStaffId(staffIdInput) {
    invalidInput = "";
    if (staffIdInput.value == "") {
        invalidInput = "Please enter your staff ID number\n";
    }
    document.getElementById(STAFF_ID_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

async function checkTheatreName(theatreNameInput) {
    invalidInput = "";
    if (theatreNameInput.value != "") {
        args = "";
        theatreIdInput = document.getElementById(THEATRE_ID);
        if (theatreIdInput != null) {
            args += theatreIdInput.name + "=" + theatreIdInput.value + "&";
        }
        args += theatreNameInput.name + "=" + theatreNameInput.value;
        invalidInput = await ajaxFunction(args, FIND_THEATRE_NAME);
    }
    document.getElementById(THEATRE_NAME_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

function checkZip(zipInput) {
    invalidInput = "";
    if (zipInput.value != "" && validateNumber(zipInput.value) != "") {
        invalidInput = "Invalid zip code\n";
    }
    document.getElementById(ZIP_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

async function checkRoomNumber(roomNumberInput) {
    invalidInput = "";
    if (roomNumberInput.value != "" && validateNumber(roomNumberInput.value) != "") {
        invalidInput = "Invalid room number\n";
    }
    if (invalidInput == "") {
        args = "";
        theatreIdInput = document.getElementById(THEATRE_ID);
        if (theatreIdInput != null) {
            args += theatreIdInput.name + "=" + theatreIdInput.value + "&";
        }
        roomIdInput = document.getElementById(ROOM_ID);
        if (roomIdInput != null) {
            args += roomIdInput.name + "=" + roomIdInput.value + "&";
        }
        args += roomNumberInput.name + "=" + roomNumberInput.value;
        invalidInput = await ajaxFunction(args, FIND_ROOM_NUMBER);
    }
    document.getElementById(ROOM_NUMBER_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

function checkSections(sectionsInput) {
    invalidInput = "";
    if (sectionsInput.value != "" && validateNumber(sectionsInput.value) != "") {
        invalidInput = "Invalid number of secions\n";
    }
    document.getElementById(SECTIONS_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

function checkSeats(seatsInput) {
    invalidInput = "";
    if (seatsInput.value != "" && validateNumber(seatsInput.value) != "") {
        invalidInput = "Invalid number of seats\n";
    }
    document.getElementById(SEATS_ERROR).innerHTML = invalidInput;
    return invalidInput;
}

function ajaxFunction(args, url) {
    return new Promise((resolve) => {
        var xhttp = new XMLHttpRequest();
        xhttp.open("POST", url);
        xhttp.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
        xhttp.send(args);
        xhttp.onreadystatechange = function () {
            if (this.readyState == 4 && this.status == 200) {
                resolve(this.responseText);
            }
        };
    });
}

function validateUserName(userName) {
    if (userName == "") return "Username field cannot be empty\n";
    if (/[^a-zA-Z0-9]/.test(userName)) return "Username must contain only letters and numbers\n";
    if (!/^[a-zA-Z]/.test(userName)) return "Username must contain a letter as first character\n";
    if (userName.length < USERNAME_MIN_LENGTH) return "Username must contain at least " + USERNAME_MIN_LENGTH + " characters\n";
    if (userName.length > USER_NAME_MAX_LENGTH) return "Username must not contain more than " + USER_NAME_MAX_LENGTH + " characters\n";
    return "";
}

function validateEmail(email) {
    if (email == "") return "Email address field cannot be empty\n";
    if (!/\S+@\S+\.\S+/.test(email)) return "Invalid email address\n";
    return "";
}

function validatePassword(password) {
    if (password == "") return "Passwords field cannot be empty\n";
    if (!/[a-z]/.test(password)) return "Passwords must contain at least a lowercase letter\n";
    if (!/[A-Z]/.test(password)) return "Passwords must contain at least an uppercase letter\n";
    if (!/[0-9]/.test(password)) return "Passwords must contain at least a number\n";
    if (password.length < PASSWORD_MIN_LENGTH) return "Passwords must be at least " + PASSWORD_MIN_LENGTH + " characters\n";
    return "";
}

function validateRePassword(password, rePassword) {
    if (rePassword == "") return "Re-enter passwords field cannot be empty\n";
    if (password != rePassword) return "Passwords are not matched\n";
    return "";
}

function validateRole(role) {
    if (role == "") {
        return "Please select a role\n";
    }
    if (!(role == "admin" || role == "manager" || role == "faculty")) {
        return "Invalid role selected\n";
    }
    return "";
}

function validateNumber(number) {
    if (!/^\d+$/.test(number)) {
        return "Invalid input number\n";
    }
    return "";
}

File: src/main/java/moviebuddy/util/Validation.java
package moviebuddy.util;

import java.time.LocalDate;
import java.time.LocalTime;
// import java.util.List;

import org.jsoup.Jsoup;
import org.jsoup.safety.Whitelist;

// import moviebuddy.model.Schedule;
// import moviebuddy.model.ShowTime;

public class Validation {
    private static final int USERNAME_MIN_LENGTH = 2;
    private static final int USERNAME_MAX_LENGTH = 20;
    private static final int PASSWORD_MIN_LENGTH = 8;
    private static final int STAFF_ID_LENGTH = 6;

    private Validation() {
    }

    public static String sanitize(String input) {
        return Jsoup.clean(input.trim(), Whitelist.none());
    }

    public static String validateSignUpForm(String userName, String email, String password, String rePassword) {
        String errorMessage = validateUserName(userName);
        if (!errorMessage.isEmpty()) {
            return errorMessage;
        }
        errorMessage = validateEmail(email);
        if (!errorMessage.isEmpty()) {
            return errorMessage;
        }
        errorMessage = validatePassword(password);
        if (!errorMessage.isEmpty()) {
            return errorMessage;
        }
        errorMessage = validateRePassword(password, rePassword);
        if (!errorMessage.isEmpty()) {
            return errorMessage;
        }
        return "";
    }

    public static String validateStaffSignUpForm(String roleInput, String theatreLocation, String userName,
            String email, String password) {
        String errorMessage = validateRole(roleInput);
        if (!errorMessage.isEmpty()) {
            return errorMessage;
        }
        errorMessage = validateTheatreLocation(theatreLocation);
        if (!roleInput.equals(S.ADMIN) && !errorMessage.isEmpty()) {
            return errorMessage;
        }
        errorMessage = validateUserName(userName);
        if (!errorMessage.isEmpty()) {
            return errorMessage;
        }
        errorMessage = validateEmail(email);
        if (!errorMessage.isEmpty()) {
            return errorMessage;
        }
        errorMessage = validatePassword(password);
        if (!errorMessage.isEmpty()) {
            return errorMessage;
        }
        return "";
    }

    public static String validateSignInForm(String email, String password) {
        if (email.isEmpty()) {
            return "Please enter your email\n";
        }
        return "";
    }

    public static String validateStaffSignInForm(String staffId, String password) {
        if (staffId.isEmpty()) {
            return "Please enter your staff ID number\n";
        }
        if (!validateStaffId(staffId).isEmpty()) {
            return "Invalid staff ID/password! Please try again\n";
        }
        return "";
    }

    public static String validateTheatreForm(String theatreName, String address, String city, String state,
            String country, String zip) {
        if (theatreName.isEmpty() || address.isEmpty() || city.isEmpty() || state.equals("none") || country.isEmpty()
                || zip.isEmpty()) {
            return "* required fields";
        }
        if (!validateNumber(zip).isEmpty()) {
            return "Invalid zip code";
        }
        return "";
    }

    public static String validateTicketPriceForm(String startTime, String price) {
        if (startTime.isEmpty()) {
            return "Please enter start time";
        }
        String errorMessage = validateTime(startTime);
        if (!errorMessage.isEmpty()) {
            return errorMessage;
        }
        if (price.isEmpty()) {
            return "Please enter price";
        }
        errorMessage = validateDouble(price);
        if (!errorMessage.isEmpty()) {
            return "Invalid price";
        }
        return "";
    }

    public static String validateRoomForm(String roomNumber, String sections, String seats) {
        if (roomNumber.isEmpty() || sections.isEmpty() || seats.isEmpty()) {
            return "* required fields";
        }
        String errorMessage = validateNumber(roomNumber);
        if (!errorMessage.isEmpty()) {
            return "Invalid room number";
        }
        errorMessage = validateNumber(sections);
        if (!errorMessage.isEmpty()) {
            return "Invalid number of secions";
        }
        errorMessage = validateNumber(seats);
        if (!errorMessage.isEmpty()) {
            return "Invalid number of seats";
        }
        return "";
    }

    public static String validateMovieForm(String title, String releaseDate, String duration, String trailer,
            String description) {
        if (title.isEmpty() || releaseDate.isEmpty() || duration.isEmpty() || trailer.isEmpty()
                || description.isEmpty()) {
            return "* required fields";
        }
        String errorMessage = validateDate(releaseDate);
        if (!errorMessage.isEmpty()) {
            return errorMessage;
        }
        if (!validateNumber(duration).isEmpty()) {
            return "Invalid duration";
        }
        return "";
    }

    public static String validateScheduleForm(String showDate, String startTime, String roomNumber) {
        if(showDate.isEmpty()){
            return "Please enter show date";
        }
        String errorMessage = validateDate(showDate);
        if(!errorMessage.isEmpty()){
            return errorMessage;
        }
        if(startTime.isEmpty()){
            return "Please enter start time";
        }
        errorMessage = validateTime(startTime);
        if(!errorMessage.isEmpty()){
            return errorMessage;
        }
        if(roomNumber.equals("none")){
            return "Please select a room";
        }
        return "";
    }

    public static String validateUserName(String userName) {
        if (userName.isEmpty()) {
            return "Username field cannot be empty\n";
        }
        if (!userName.matches("(.*[a-zA-Z0-9].*)")) {
            return "Username must contain only letters and numbers\n";
        }
        if (!userName.matches("(^[a-zA-Z].*)")) {
            return "Username must a letter as first character\n";
        }
        if (userName.length() < USERNAME_MIN_LENGTH) {
            return "Username must contain at least " + USERNAME_MIN_LENGTH + " characters\n";
        }
        if (userName.length() > USERNAME_MAX_LENGTH) {
            return "Username must not contain more than " + USERNAME_MAX_LENGTH + " characters\n";
        }
        return "";
    }

    public static String validateEmail(String email) {
        if (email.isEmpty()) {
            return "Email address field cannot be empty\n";
        }
        if (!email.matches("^[\\w-_\\.+]*[\\w-_\\.]\\@([\\w]+\\.)+[\\w]+[\\w]$")) {
            return "Invalid email address\n";
        }
        return "";
    }

    public static String validatePassword(String password) {
        if (password.isEmpty()) {
            return "Passwords field cannot be empty\n";
        }
        if (!password.matches("(.*[a-z].*)")) {
            return "Passwords must contain at least a lowercase letter\n";
        }
        if (!password.matches("(.*[A-Z].*)")) {
            return "Passwords must contain at least an uppercase letter\n";
        }
        if (!password.matches("(.*[0-9].*)")) {
            return "Passwords must contain at least a number\n";
        }
        if (password.length() < PASSWORD_MIN_LENGTH) {
            return "Passwords must be at least " + PASSWORD_MIN_LENGTH + " characters\n";
        }
        return "";
    }

    public static String validateRePassword(String password, String rePassword) {
        if (rePassword.isEmpty()) {
            return "Re-enter passwords field cannot be empty\n";
        }
        if (!password.equals(rePassword)) {
            return "Passwords are not matched\n";
        }
        return "";
    }

    public static String validateRole(String role) {
        if (role.equals("none")) {
            return "Please select a role\n";
        }
        if (!(role.equals("admin") || role.equals("manager") || role.equals("faculty"))) {
            return "Invalid role\n";
        }
        return "";
    }

    public static String validateTheatreLocation(String location) {
        if (location.equals("none")) {
            return "Please select a theatre location\n";
        }
        return "";
    }

    public static String validateStaffId(String staffId) {
        if (staffId.length() != STAFF_ID_LENGTH || !staffId.matches("[0-9]+")) {
            return "Incorrect staff ID number format\n";
        }
        return "";
    }

    public static String validateNumber(String number) {
        if (!number.matches("[0-9]+")) {
            return "Invalid number input\n";
        }
        return "";
    }

    public static String validateDouble(String number) {
        try {
            Double.parseDouble(number);
        } catch (Exception e) {
            return "Invalid double input\n";
        }
        return "";
    }

    public static String validateDate(String date) {
        try {
            LocalDate.parse(date);
        } catch (Exception e) {
            return "Invalid date input\n";
        }
        return "";
    }

    public static String validateTime(String time) {
        try {
            LocalTime.parse(time);
        } catch (Exception e) {
            return "Invalid time input\n";
        }
        return "";
    }

    // public static String checkScheduleConflict(List<Schedule> schedule, ShowTime
    // interval) {
    // for (Schedule s : schedule) {
    // ShowTime st = s.getShowTime();
    // if (interval.isConflict(st)) {
    // String message = String.format("Time conflict - Schedule ID: %s on %s at
    // %s-%s room: %s", s.getScheduleId(),
    // s.displayShowDate(), st.getStartTime(), st.getEndTime(), s.getRoomNumber());
    // return message;
    // }
    // }
    // return "";
    // }
}
